Org owners can be excluded from auto-review

asvoboda opened the pull request.

asvoboda committed:
Org owners can be excluded from auto-review

asvoboda committed:
fix leaf node tests

bluekeyes commented:
Can you explain the background of this more? My understanding was that we only wanted to exclude organization admins if the rule requires admin approval _and_ they are not admins of the repository either directly or via a team.

As currently implemented, I believe this will remove org admins all the time, even if they are also admins on the repo through a team or directly. In our internal case, all the infrastructure team members are all organization admins, but removing them from review selection feels counter to the initial feature request.

asvoboda commented:
You're right about the goal to exclude org admins if they aren't admins of the repo directly or by team.

asvoboda committed:
Refactor option to AdminScope; add more tests

asvoboda committed:
Better collaborator differentiation

asvoboda committed:
Add parsing test

asvoboda commented on policy/common/result.go:
```diff
@@ -37,13 +37,22 @@ func (s EvaluationStatus) String() string {
 
+type AdminScope string
+
+const (
```
Should we bail if an option provided doesn't match here?

bluekeyes replied:
I think that's best, but I'm not sure how easy it is to implement. Maybe if `AdminScope` implemented the `MarshalText`/`UnmarshalText` methods, the YAML parsing would fail automatically? I don't remember if there's a separate validation step after parsing.

We'll also need to decide how the default (empty string) is handled.

asvoboda replied:
I think a sane default might be to assume `org` if unset? If you require admins to +1, but otherwise don't hint about what kind of admin, we should probably pick from anyone.

bluekeyes commented on README.md:
```diff
@@ -203,6 +203,8 @@ options:
     # False by default
     enabled: true
+    # Include organization owners in automatic review. False by default.
+    org_owner: false
```
Update this for the new `admin_scope` property

bluekeyes commented on policy/common/result.go:
```diff
@@ -37,13 +37,22 @@ func (s EvaluationStatus) String() string {
+const (
+	User AdminScope = "user"
+	Team AdminScope = "team"
+	Org  AdminScope = "org"
```
style: let's call these `ScopeUser` or `AdminScopeUser`... seeing `common.User` in another file could mean a lot of things.

bluekeyes commented on policy/reviewer/reviewer.go:
```diff
@@ -34,6 +34,9 @@ func findLeafChildren(result common.Result) []common.Result {
 		}
 	} else {
 		for _, c := range result.Children {
+			if c == nil {
```
When are children `nil`? Is that a result of this change or was that always the case?

asvoboda replied:
I changed this because I was inserting a nil `Result` in the tests

bluekeyes replied:
Would it possible to not use `nil` children in the tests if this is not a case that occurs in the application? If children can legitimately be `nil` then I believe there are other places that need to be updated.

asvoboda replied:
Fair point. I'll fix up the tests and remove this.

bluekeyes commented on pull/context.go:
```diff
@@ -36,6 +36,9 @@ type MembershipContext interface {
 	// OrganizationMembers returns the list of usernames in the given organization.
 	OrganizationMembers(org string) ([]string, error)
+
+	// OrganizationMembers returns the list of owner usernames in the given organization.
```
`OrganizationOwners`

bluekeyes commented on policy/reviewer/reviewer.go:
```diff
@@ -124,31 +127,94 @@ func FindRandomRequesters(ctx context.Context, prctx pull.Context, result common
 
-		allCollaboratorPermissions, err := prctx.RepositoryCollaborators()
+		allCollaboratorPermissions := make(map[string]string)
+		allCollaboratorAndPermissions, err := prctx.RepositoryCollaborators()
```
The distinction between these variables is unclear... what does the extra `And` in the second one mean for the functionality?

asvoboda replied:
I was trying to come up with a better name here. The idea is that we want to only consider some collaborators, and not all, when checking which admins can be auto-assigned. I've tried to rename these to `allCollaborators` and `collaboratorsToConsider` to be a little more clear.

bluekeyes commented on policy/reviewer/reviewer.go:
```diff
@@ -124,31 +127,94 @@ func FindRandomRequesters(ctx context.Context, prctx pull.Context, result common
+			// When not looking for admins, we want to check with everyone
+			allCollaboratorPermissions = allCollaboratorAndPermissions
+		} else {
+			// Determine what the scope of requested admins should be
```
This feels significant enough to merit a separate helper function, maybe `selectAdmins` to match `selectTeamMembers` and `selectOrgMembers`.

bluekeyes commented on policy/reviewer/reviewer.go:
```diff
@@ -124,31 +127,94 @@ func FindRandomRequesters(ctx context.Context, prctx pull.Context, result common
+				logger.Debug().Msg("Selecting admin users with direct collaboration rights")
+				directCollaborators, err := prctx.DirectRepositoryCollaborators()
+				if err != nil {
+					logger.Error().Err(err).Msgf("Unable to get list of direct collaborators on %s", prctx.RepositoryName())
```
Seeing these error logs that continue on with execution is surprising (even if it works); I think this is another reason to make a helper function.

bluekeyes commented on policy/reviewer/reviewer.go:
```diff
@@ -124,31 +127,94 @@ func FindRandomRequesters(ctx context.Context, prctx pull.Context, result common
+					allCollaboratorPermissions[a] = common.GithubAdminPermission
+				}
+
+				break
```
style: `break` is unnecessary in Go unless terminating the `case` early

bluekeyes commented on policy/reviewer/reviewer.go:
```diff
@@ -124,31 +127,94 @@ func FindRandomRequesters(ctx context.Context, prctx pull.Context, result common
+				}
+
+				for _, o := range orgOwners {
+					allUsers[o] = struct{}{}
```
Why do the other cases use `shoveInToMap`, but this adds the users directly in a single loop? I think the single loop version (this one) is preferable, but would rather have consistency.

asvoboda replied:
I'll remove the `shoveIntoMap` calls

bluekeyes commented on pull/github.go:
```diff
@@ -289,6 +289,30 @@ func (ghc *GitHubContext) HasReveiwers() (bool, error) {
+		opt.Page = resp.NextPage
+	}
+
+	return allTeams, nil
```
Should this be cached like other list output in the context?

asvoboda replied:
We should get v3 caching by default, but I'm open to also caching it more explicitly here.

bluekeyes replied:
Ok, we can try it with only the v3 caching for now.

bluekeyes commented on pull/github.go:
```diff
@@ -356,20 +380,36 @@ func (ghc *GitHubContext) loadCollaboratorData() error {
+		if ghc.collaborators != nil {
+			return ghc.collaborators, nil
+		}
+	}
```
Instead of implementing this switching in this function, what if `loadCollaborators` only returned the map of collaborators and the callers were responsible for checking and initializing the cache?

That would also let you replace the `direct` parameter and take the `affiliation` value directly.

asvoboda replied:
I felt like it was a little more clunky if the clients need to worry about the caching, but I'm open to changing it if you feel like this is a worse outcome.

bluekeyes commented on pull/github_membership.go:
```diff
@@ -130,8 +130,9 @@ func (mc *GitHubMembershipContext) IsCollaborator(org, repo, user, desiredPerm s
 }
 
-func (mc *GitHubMembershipContext) OrganizationMembers(org string) ([]string, error) {
+func (mc *GitHubMembershipContext) organizationMembers(org, perm string) ([]string, error) {
```
style: `role` instead of `perm`

bluekeyes commented on pull/pulltest/context.go:
```diff
@@ -171,6 +190,14 @@ func (c *Context) OrganizationMembers(org string) ([]string, error) {
+		return nil, c.OrgOwnerError
+	}
+
+	return c.OrgOwners, nil
```
style: simplify to `return c.OrgOwners, c.OrgOwnerError`

bluekeyes reviewed (commented):
When I originally proposed the `AdminScope` idea, I was thinking it would be cumulative, so that the `team` scope could also select users, and the `org` scope could also select users and teams. Do you think this is useful? Will people using admin rules be satisfied with picking a single type of admin for review requests? If you do want to implement this, I think it's relatively simple: invert the order of the `switch` and add `fallthrough` at the end of each `case`.

And unrelated to your code, it's really annoying how the GitHub API let's you filter on things like affiliation but provides no fields in the result object that would let client code do the same filtering. It would be a lot nicer to combine some of these query methods and return an object with the users grouped however we wanted.

asvoboda committed:
fixup style nits

asvoboda committed:
Refactor out selection of admins

asvoboda committed:
Update readme

asvoboda committed:
remove shoveIntoMap

iamdanfox commented on README.md:
```diff
@@ -203,6 +203,11 @@ options:
+    # Admin of a repo is granted in three ways: direct user admin on the Repository, team admin on the repository,
+    # or as an Organization Owner.
+    # To reflect these permission sources, admin_scope can be set to (user|team|org)
+    admin_scope: "org"
```
Are you sure you want to commit to this configurability? Seems fine to keep the settings internally, but I'd be tempted to just hard-code this setting so that only people who have admin perms via the `user` or `team` scope get requested (and org level admins never get requested).

asvoboda replied:
Fair q. I think in general we prefer to have policy bot accept a variety of configurations, and for users to make their own decisions about how they want to achieve the best pull review ux. I don't think its super great to hard code assumptions about how to best review things into the bot.

This feature is actually pretty simple if I could figure out the right incantation for https://developer.github.com/v4/object/permissionsource/, but it appears to always return an empty list for github apps.

asvoboda replied:
The more I think about it, the more I think you're right. We don't need to expose the complexity today, and we can always make it more configurable later. I'll push up a commit to reflect this.

asvoboda committed:
Remove admin_scope concept

asvoboda committed:
simplify selectAdmins func

bluekeyes commented on policy/reviewer/reviewer.go:
```diff
@@ -97,6 +94,28 @@ func selectOrgMembers(prctx pull.Context, allOrgs []string, r *rand.Rand) ([]str
+	var adminUsers []string
+	teams, err := prctx.Teams()
+	if err != nil {
+		return nil, errors.Wrap(err, "Unable to get list of teams collaborators")
```
style: almost all other `errors.Wrap` calls in this project use a message like `failed to ...` - can you update these to match?

bluekeyes commented on policy/reviewer/reviewer.go:
```diff
@@ -106,54 +125,70 @@ func FindRandomRequesters(ctx context.Context, prctx pull.Context, result common
 		allUsers := make(map[string]struct{})
-		shoveIntoMap(allUsers, child.ReviewRequestRule.Users)
+
+		if len(child.ReviewRequestRule.Users) > 0 {
```
style: this check is redundant, `range` on an empty loop won't do anything

bluekeyes commented on policy/reviewer/reviewer.go:
```diff
@@ -106,54 +125,70 @@ func FindRandomRequesters(ctx context.Context, prctx pull.Context, result common
+		// everyone, which includes org admins
+		if !child.ReviewRequestRule.Admins {
+			// When not looking for admins, we want to check with all possible collaborators
+			collaboratorsToConsider = allCollaborators
```
These names are better now, but seeing it in the new limited context makes me think the distinction is not necessary at all, and is actually incorrect.

The `allUsers` set is what we actually use to select reviewers, while `collboratorsToConsider` is only used to figure out which elements of `allUsers` are collaborators. In the admin case, we add the team admins to `allUsers` but then we additionally limit the set of collaborators to admins as well. When an approval rule allows both admins and other conditions, I think this artificially limits the set of reviewers to only the admins. And when the rule only allows admins, the admin/not-admin filtering is performed by `selectAdmins`, so the additional filtering at this level is redundant.

Am I missing something?

asvoboda replied:
I think you're right. I added the distinction previously when trying to find the set of admins and then remove certain users like org owners. Now that we're going with a purely additive approach, we can simplify and remove this bug.

bluekeyes reviewed (commented):
Two logic questions, two small style things, and then I think this is good to go.

asvoboda committed:
Fixup style comments

asvoboda committed:
Remove distinct around collaborators and admins

asvoboda committed:
Select users more evenly

Instead of selecting one team, find all members for all teams
and select from the unique set

bluekeyes commented on policy/reviewer/reviewer.go:
```diff
@@ -73,96 +73,130 @@ func selectRandomUsers(n int, users []string, r *rand.Rand) []string {
+			fullTeamName := fmt.Sprintf("%s/%s", prctx.RepositoryOwner(), team)
+			admins, err := prctx.TeamMembers(fullTeamName)
+			if err != nil {
+				return nil, errors.Wrapf(err, "Unable to get list of members for %s", team)
```
style: one missed `failed to`

bluekeyes commented on policy/reviewer/reviewer.go:
```diff
@@ -73,96 +73,130 @@ func selectRandomUsers(n int, users []string, r *rand.Rand) []string {
+
+			for _, admin := range admins {
+				allUsers[admin] = struct{}{}
+				collaboratorsToConsider[admin] = common.GithubAdminPermission
```
I thought these users would already be in `collaboratorsToConsider` as part of listing all collaborators. Is that not true?

asvoboda replied:
you're correct, I missed this in the last commit.

asvoboda committed:
style nits

iamdanfox commented:
Awesome :) lmk when we can try this out on some of our oss/internal repos 🎉🎉
